### ISSUE TITLE
Implement Domain-Aware Embedding Service via LLM Distillation

### ISSUE THREAD

**Issue by yossiovadia** [vllm-project/semantic-router#1131]
## Overview

Implement domain-aware embedding service for semantic-router based on the paper [Adaptation of Embedding Models to Financial Filings via LLM Distillation](https://arxiv.org/pdf/2512.08088).

**Goal**: Provide an embedding API (like OpenAI's) with competitive advantages:
1. **Domain-aware embeddings** - Specialized for specific domains via iterative hard-negative mining
2. **Long context support** - Leverage existing models with 32K context

## Business Value

Enable semantic-router to serve as an embedding provider for RAG systems and agents, competing with OpenAI's embedding API through domain specialization.

---

## ✅ POC Complete: Medical Domain

**Result: +71.18% MRR@5 improvement** (exceeded +15-25% target)

| Metric | Baseline | After Training | Improvement |
|--------|----------|----------------|-------------|
| **MRR@5** | 0.4354 | 0.7453 | **+71.18%** |
| **Recall@5** | 0.4896 | 0.6900 | **+40.92%** |

### Pre-trained Model

The trained medical embedding model is available on HuggingFace:
**[llm-semantic-router/mmbert-embed-medical](https://huggingface.co/llm-semantic-router/mmbert-embed-medical)**

```python
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("llm-semantic-router/mmbert-embed-medical", trust_remote_code=True)
embeddings = model.encode(["What are the symptoms of diabetes?"])
```

### Implementation: `src/training/domain_adapted_embeddings/`

| File | Purpose |
|------|---------|
| `prepare_data.py` | Converts Q&A datasets to training format |
| `train.py` | Iterative hard-negative mining + fine-tuning |
| `README.md` | Overview and results |
| `USAGE.md` | Training instructions and code examples |

### How It Works

**The Core Idea**: Fine-tune an embedding model so questions and their correct answers have closer vector representations.

**Process**:
1. **Encode** all questions and answers into vectors using current model
2. **Rank** answers for each question by similarity
3. **Find hard cases**:
   - Hard positives: correct answers ranked too low
   - Hard negatives: wrong answers ranked too high
4. **Create triplets**: (question, correct_answer, wrong_answer)
5. **Train**: Adjust model to pull correct answers closer, push wrong answers away
6. **Repeat**: Mine new hard cases with improved model, accumulate triplets

```
Iteration 1: Mine → Train → Evaluate (MRR: 0.4354 → 0.6528, +49.93%)
Iteration 2: Mine → Accumulate → Train → Evaluate (MRR: 0.6528 → 0.7453, +71.18%)
```

### Adaptation from Paper

| Paper Approach | Our Implementation | Why |
|----------------|-------------------|-----|
| LLM-as-judge scores relevance (1-4) | Ground-truth labels | We had labeled Q&A data, no LLM needed |
| Generate queries from corpus | Use existing questions | MedQuAD provides questions |
| Chunk large PDFs | Use answers directly | Q&A answers are already reasonable size |

The paper's core contribution (iterative hard-negative mining) is preserved. We simplified by using ground-truth labels instead of LLM scoring — this works when you have labeled Q&A data.

### Critical Hyperparameters

| Parameter | Value | Why It Matters |
|-----------|-------|----------------|
| Learning rate | 5e-5 | Lower values (5e-7) result in minimal improvement |
| Margin | 0.1 | Default (~5.0) causes severe forgetting |
| Easy:Hard ratio | 2:1 | Easy triplets reinforce existing knowledge |
| Accumulate triplets | Yes | Don't replace, add each iteration |

### Training Configuration

| Parameter | Value |
|-----------|-------|
| Base model | llm-semantic-router/mmbert-embed-32k-2d-matryoshka |
| Dataset | MedQuAD (13,125 training queries) |
| Iterations | 2 |
| Epochs per iteration | 2 |
| Batch size | 8 |
| GPU | NVIDIA L4 (24GB) |

---

## Dataset Format

Any Q&A dataset works. Format:

```json
[
  {"question": "What are the symptoms of diabetes?", "answer": "Diabetes symptoms include..."},
  {"question": "How is hypertension treated?", "answer": "Hypertension treatment involves..."}
]
```

Or use HuggingFace datasets directly:
```bash
python prepare_data.py --source huggingface --dataset keivalya/MedQuad-MedicalQnADataset
```

---

## Quick Start

```bash
cd src/training/domain_adapted_embeddings

# 1. Install dependencies
pip install -r requirements.txt

# 2. Prepare data
python prepare_data.py --source huggingface --dataset keivalya/MedQuad-MedicalQnADataset

# 3. Train
python train.py --data-dir data --output-dir models/trained

# 4. Use trained model
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("models/trained/best", trust_remote_code=True)
embeddings = model.encode(["Your query here"])
```

---

## Limitation: Multi-Domain Integration (TBD)

Each domain-adapted model is fine-tuned independently. When multiple domains exist (medical, legal, finance, etc.), a strategy is needed to integrate them into the semantic router:

| Option | Approach | Trade-off |
|--------|----------|-----------|
| **Multi-domain fine-tuning** | Train one model on all domains together | Simple deployment, but domains may interfere |
| **Two-stage routing** | Domain classifier → domain-specific embeddings | Best accuracy, more complexity |
| **Ensemble** | Run through multiple models, combine scores | Handles cross-domain queries, higher latency |

Currently, this pipeline produces standalone domain models. Integration with the semantic router's existing domain signal infrastructure is future work.

---

## Phase 2: Production Integration (Future)

After POC proves approach works:
1. Load specialized model in Rust via candle
2. FFI bindings for Go
3. Configuration support for domain selection
4. Model registry for multiple domains

---

## References

- Paper: https://arxiv.org/pdf/2512.08088
- Base Model: [llm-semantic-router/mmbert-embed-32k-2d-matryoshka](https://huggingface.co/llm-semantic-router/mmbert-embed-32k-2d-matryoshka)
- Pre-trained Medical Model: [llm-semantic-router/mmbert-embed-medical](https://huggingface.co/llm-semantic-router/mmbert-embed-medical)
- Dataset: [MedQuAD](https://huggingface.co/datasets/keivalya/MedQuad-MedicalQnADataset)
